### PR TITLE
feat: add support for the namespace file resource and datasource

### LINKED ISSUE #91: Add a Namespace File resource
### Feature description

Allow deploying files for access from the VS Code editor from Terraform.

### PR DESCRIPTION
**Pull request by moadibfr**
closes #91

### CODE REVIEW

File: docs/data-sources/namespace_file.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+---
+# generated by https://github.com/hashicorp/terraform-plugin-docs
+page_title: "kestra_namespace_file Data Source - terraform-provider-kestra"
+subcategory: ""
+description: |-
+  Use this data source to access information about an existing Namespace File
```
Comment (anna-geller) [inline docs/data-sources/namespace_file.md:6]:
```suggestion
  Use this data source to manage Namespace File resource
```

File: docs/data-sources/namespace_file.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+---
+
+# kestra_namespace_file (Data Source)
+
+Use this data source to access information about an existing Namespace File
```
Comment (anna-geller) [inline docs/data-sources/namespace_file.md:11]:
```suggestion
Use this data source to manage the Namespace File resource
```

File: docs/data-sources/namespace_file.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+<!-- schema generated by tfplugindocs -->
+## Schema
+
+### Required
+
+- `filename` (String) The path to the namespace file that will be created.
+Missing parent directories will be created.
+If the file already exists, it will be overridden with the given content.
+- `namespace` (String) The Namespace file namespace.
```
Comment (anna-geller) [inline docs/data-sources/namespace_file.md:30]:
```suggestion
- `namespace` (String) The namespace to which the namespace file resource should be deployed
```

File: docs/resources/namespace_file.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  content   = <<EOT
+#!/bin/bash
+echo "Hello World"
+EOT
+}
+
+resource "kestra_namespace_file" "withsource" {
+  namespace = "io.kestra.mynamespace"
+  filename  = "/path/my-file.sh"
+  content   = "./kestra/file.sh"
```
Comment (anna-geller) [inline docs/resources/namespace_file.md:28]: interesting, so there's no need to read the file's content with terraform functions? you read it by path? and if so, why is this path different from the one above i.e. why not:
```
  filename  = "/path/my-file.sh"
  content   = "/path/my-file.sh"
```
(just to confirm)

Comment (anna-geller) [inline docs/resources/namespace_file.md:28]:
```suggestion
  content   = "bash ./kestra/file.sh"
```
super optional to make it clear it's file's content and not a path :) I initially thought this was another path and got confused, while this is a totally valid file's content.

File: examples/data-sources/kestra_namespace_file/data-source.tf [hunk 5]
```diff
@@ -0,0 +1,4 @@
+data "kestra_namespace_file" "example" {
+  namespace_ = "io.kestra.mynamespace"
+  filename   = "my-file.yml"
```
Comment (anna-geller) [inline examples/data-sources/kestra_namespace_file/data-source.tf:3]:
```suggestion
  filename   = "myscript.py"
```
optional, to show that this is meant more for script deployments than for flow deployments

File: examples/resources/kestra_namespace_file/resource.tf [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+  content   = <<EOT
+#!/bin/bash
+echo "Hello World"
+EOT
+}
+
+resource "kestra_namespace_file" "withsource" {
+  namespace = "io.kestra.mynamespace"
+  filename  = "/path/my-file.sh"
+  content   = "./kestra/file.sh"
```
Comment (anna-geller) [inline examples/resources/kestra_namespace_file/resource.tf:13]:
```suggestion
```
maybe skip this time? to give one example showing how to deploy based on local path without specifying file's content inline in the terraform file

File: internal/provider/resource_namespace_file.go [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+		DeleteContext: resourceNamespaceFileDelete,
+		Schema: map[string]*schema.Schema{
+			"tenant_id": {
+				Description: "The tenant id.",
+				Type:        schema.TypeString,
+				Optional:    true,
+				ForceNew:    true,
+			},
+			"namespace": {
+				Description: "The Namespace file namespace.",
```
Comment (anna-geller) [inline internal/provider/resource_namespace_file.go:31]:
```suggestion
				Description: "The namespace of the namespace file.",
```
seems a little clearer

File: internal/provider/resource_namespace_file.go [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+			},
+			"filename": {
+				Description: "The path to the namespace file that will be created.\n" +
```
Comment (anna-geller) [inline internal/provider/resource_namespace_file.go:37]:
```suggestion
				Description: "The local path to the namespace file that will be deployed.\n" +
```
to show it's a path on a local machine, not on kestra

I often find it confusing in some CLIs so worth clarifying :)

Comment (anna-geller) [reply internal/provider/resource_namespace_file.go:37]: oh wait, this is the target path on kestra VS Code? and the source is the local path?

if so, perhaps worth changing the naming to source_path and target_path, or local_path + kestra_path?

**Review by anna-geller — APPROVED** [review]
Super flexible. I'd only recommend considering clarifying the paths more, as I got confused about which path is meant as the source and which as the destination.

Perhaps using `source_path` and `destination_path`?

**Comment by moadibfr** [comment]
I think using the file built-in might cause issues with binary (non UTF-8) files that's mainly why I allowed to input a real path or a string content like https://registry.terraform.io/providers/hashicorp/local/latest/docs/resources/file does. 🤷

**Comment by anna-geller** [comment]
Thanks so much for the info, Martin 👍

**Pull request merged**